### ISSUE TITLE
"ItemNotFound" when uploading file

### ISSUE THREAD

**Issue by tipa** [microsoftgraph/msgraph-sdk-dotnet#385]
using Nuget package 1.12.0 in my UWP app

 ### Expected behavior
I am able to upload a file to OneDrive (into the apps app folder).

### Actual behavior
An excpetion is thrown (see below) - but only sometimes!
I feel like it mostly (or only) happens the first time the app tries to upload something to its (newly created) app folder. It usually (if not always) succeeds on a second attempt.

> Microsoft.Graph.ServiceException: Code: itemNotFoundMessage: Item does not existInner error   at Microsoft.Graph.HttpProvider.<SendAsync>d__19.MoveNext() + 0x5ac--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x21   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x5c   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task) + 0x44   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task) + 0x1c   at Microsoft.Graph.BaseRequest.<SendRequestAsync>d__36.MoveNext() + 0x475--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x21   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x5c   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task) + 0x44   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task) + 0x1c   at Microsoft.Graph.BaseRequest.<SendAsync>d__32`1.MoveNext() + 0x12f--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x21   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x5c   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task) + 0x44   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task) + 0x1c   at Diarium.OneDriveHelper.<UploadFile>d__15.MoveNext() + 0x2ad--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x21   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x5c   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task) + 0x44   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task) + 0x1c   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult() + 0xb   at Diarium.OneDriveHelper.<Sync>d__1.MoveNext() + 0x150e

### Steps to reproduce the behavior
```
static async Task UploadFile(GraphServiceClient graphClient, string filePath, MemoryStream stream)
{
    if (stream.Length > 4194304)
    {
        var session = await graphClient.Drive.Special.AppRoot.ItemWithPath(filePath).CreateUploadSession().Request().PostAsync();
        await new ChunkedUploadProvider(session, graphClient, stream).UploadAsync();
    }
    else
    {
        await graphClient.Drive.Special.AppRoot.ItemWithPath(filePath).Content.Request().PutAsync<DriveItem>(stream);
    }
}
```

Note: I am using the "beta" version of the API:
```
var client = new GraphServiceClient("https://graph.microsoft.com/beta", new DelegateAuthenticationProvider(reqMsg =>
{
    reqMsg.Headers.Authorization = new AuthenticationHeaderValue("bearer", authResult.AccessToken);
    return Task.CompletedTask;
}));
```

**Comment by tipa** [comment]
Any news? This issue is still present for some users.

**Comment by MIchaelMainer** [comment]
You mentioned that this fails after creating a new folder. How long after the folder is created do you attempt to upload an item to it? I've found in the past that I've had to delay some requests that target OneDrive as there is latency between when the call to create an item (or folder) completes and being able to access it. 

I suggest that you do one of the following after you create the folder:
* Make a call to the folder to make sure it is accessible to the API before uploading content to it. 
* Retry UploadFile on the itemNotFound code. 
* Add a delay - I don't like this idea, just adding it for completeness.

**Comment by tipa** [comment]
Yes, I was able to diminish the issue for nearly all users by waiting 1-2 secs after initial login before creating new files.
However, today a user had recurring problems with this error, even after multiple tries which were several minutes apart. Maybe it was just a OneDrive hickup as for 99.9% of the users, the delay seems to work.

**Comment by darrelmiller** [comment]
@tipa It is great to hear that you were able to minimize the issue.  I believe we may be able to help here.  Also, there are some fairly major outages happening across Azure and M365 today, so issues you see today may be service related.

**Comment by tipa** [comment]
Another user reports having that issue repeatedly. Anything I can ask him to do in order to debug the issue?

**Comment by MIchaelMainer** [comment]
@tipa Are you able to repro? What is different (if any) about this user's use case? Is the user using the same UWP app as the other users?

**Comment by tipa** [comment]
He uses the exact same app as my other users. As I said, it works fine for 99% of my users. Unfortunately, I am not able to reproduce personally :(
I'll put in the stack traces again as they looks slightly different than the original one:

Xamarin.Android:
`Microsoft.Graph.ServiceException: Code: itemNotFound Message: Item does not exist Inner error at Microsoft.Graph.HttpProvider.SendAsync (System.Net.Http.HttpRequestMessage request, System.Net.Http.HttpCompletionOption completionOption, System.Threading.CancellationToken cancellationToken) [0x002d1] in <76600a1e64c94b95bee3341281463404>:0 at Microsoft.Graph.BaseRequest.SendRequestAsync (System.Object serializableObject, System.Threading.CancellationToken cancellationToken, System.Net.Http.HttpCompletionOption completionOption) [0x001f1] in <76600a1e64c94b95bee3341281463404>:0 at Microsoft.Graph.BaseRequest.SendAsync[T] (System.Object serializableObject, System.Threading.CancellationToken cancellationToken, System.Net.Http.HttpCompletionOption completionOption) [0x0008b] in <76600a1e64c94b95bee3341281463404>:0 
`

Windows UWP:
```
Microsoft.Graph.ServiceException: Code: itemNotFound
Message: Item does not exist

Inner error

at Microsoft.Graph.HttpProvider.<SendAsync>d__19.MoveNext() + 0x5ac
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x21
at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x70
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task) + 0x38
at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task) + 0x17
at Microsoft.Graph.BaseRequest.<SendRequestAsync>d__36.MoveNext() + 0x4e2
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x21
at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x70
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task) + 0x38
at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task) + 0x17
at Microsoft.Graph.BaseRequest.<SendAsync>d__32`1.MoveNext() + 0x12f
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x21
at System.Runtime.Compi
```

Both methods attempt do upload a file into a subfolder of the OneDrive app folder, with permission "Files.ReadWrite.AppFolder". It is possible, that none of these folders exist at the time of uploading. Also, I am using baseUrl "https://graph.microsoft.com/beta"

**Comment by tipa** [comment]
Had another user report this error today

**Comment by MIchaelMainer** [comment]
When this happens, can you verify that the folder exists with both the API and the OneDrive/SharePoint UI?

**Comment by tipa** [comment]
I can ask my users to let me know if my app created any folder in OneDrive. Last user didn't respond though as he simply switched to the Google Drive sync that I offer and which works without problems

**Comment by MIchaelMainer** [comment]
@ificator Do you have wisdom to share on this scenario?

**Comment by ificator** [comment]
Without the `request-id` and `date` from the error in the response body it'll be hard to see what's going on. Also, the code switches between session based uploads and direct put - do we know if this happens for both of them or just one?

**Comment by tipa** [comment]
I would say usually the first file that is being uploaded from my is below 4MB so the else block is being executed - but I can ask the next user that reports that problem. Anything else that I should ask them then besides checking if any folder has been created in their OneDrive?
How can I get hold of the `request-id` when that exception is thrown? Maybe I can log it and send an event to AppCenter.

**Comment by ificator** [comment]
The error responses from the Graph service will look something like this:

```
{
    "error": {
        "code": "itemNotFound",
        "message": "Item does not exist",
        "innerError": {
            "request-id": "cdbbe332-06cf-4fa8-86a2-30afd36367b1",
            "date": "2019-05-24T15:11:38"
        }
    }
}
```

I'm not really sure how that data is manifested in the SDK though... @MIchaelMainer are the `request-id` and `date` properties in the `innerError` populated in the exception somewhere?

**Comment by MIchaelMainer** [comment]
The requestId is on the ServiceException.Error.InnerError.AdditionalData[request-id]

**Comment by tipa** [comment]
Thanks - I added those infos to my logging code and will let you know as soon as I see this error show up again

**Comment by tipa** [comment]
I now have logged a bunch occurences of this error, I'll post here a few:
request-id: ab53bdf0-c985-4b31-b045-a023190f9fc8, date: 29-5-2019 18:31:00
request-id: 0006de86-4576-47f6-b914-b0bbae7fdddc, date: 03/06/2019 15:20:43
request-id: 08b93382-3ecc-462d-9c27-f98ffe35a414, date: 03/06/2019 12:20:24
request-id: 26331dc3-bd15-4ccf-8763-e9ac12ae1a2e, date: 02/06/2019 17:28:21

Dates may be in local time, forgot to transfor it to UTC before logging. Let me know if you need any additional info

**Comment by tipa** [comment]
Any updates on this? I can provide more request-id's if needed.
As far as I can see in AppCenter, this happens mostly in the Windows UWP app, not in the Android app and only for very few users in the iOS app

**Comment by ificator** [comment]
Hi @tipa - I experimented a little and did come up with something that is plausible. Is it possible that you're attempting to create multiple files in the approot at the same time?

**Comment by tipa** [comment]
Yes, that is the case - I am sending two PUT-requests simultaneously

**Comment by ificator** [comment]
Thanks for confirming - it looks like there's a race between the creation of the parent folders for both requests. I'll see what we can do to insulate the caller from this.

**Comment by tipa** [comment]
I updated my app, removing the concurrent uploads (and execute them one after another) when the app knows that the parent folder already exist, however, the error still shows up.
request-id: 19c3d1c1-c426-4bf5-a26e-3e5f0b883e55, date: 21.06.2019 07:15:05
request-id: 1c0d5295-431b-4597-89d2-094893c227d9, date: 21.06.2019 09:15:02
request-id: 425a9d48-2e25-4f87-bfba-6c9340668b39, date: 21.06.2019 06:27:45
request-id: 50f95c9f-e8bb-4373-869c-4fe420a76985, date: 21.06.2019 06:29:05
request-id: 5682a0ba-e7fe-4eef-a7e2-b73795df3bee, date: 21.06.2019 06:59:05

**Comment by tipa** [comment]
Any updates on this one?

**Comment by tipa** [comment]
This issue is nearly half a year old now and I am still getting these errors even though I don't see me doing anything wrong in my code. Can I expect this being fixed at some point? I am happy to share additional information / request-id's if needed

**Comment by darrelmiller** [comment]
@tipa I would recommend creating a uservoice issue for this as that will get more visibility from the folks who prioritize work to be done.  https://microsoftgraph.uservoice.com/forums/920506-microsoft-graph-feature-requests

**Comment by tipa** [comment]
I created that issue as you suggested [here](https://microsoftgraph.uservoice.com/forums/920506-microsoft-graph-feature-requests/suggestions/38356555-fix-itemnotfound-when-uploading-file) 2 weeks ago.
Unfortunately (but expected) there was no reaction since then.
It saddens me that noone seems to be responsible fixing this apparent bug.

**Comment by arsnyder16** [comment]
@darrelmiller  @tipa  

I am able to reproduce a similar issue that seems unrelated to concurrent requests. If one of my users shares a file from their one drive to another user with write permissions, then anything time the uploadUrl is used they receive a 404 "ItemNotFound". This includes trying to delete the uploadsession. 

I haven't tested this across tenants, All the issues we see are two users sharing files between their onedrives within the same tenant.  

If using the standard upload for small files <4MB these scenarios work fine. If the user shares the folder that the file is contained then the upload sessions seem to work as expected.

My suspicion is that an upload session is simply creating a temporary file to upload to, but when that temporary file is created the permissions are not transferred from the original file, hence being invisible(404) to all users other than the user "owning" the drive. 

I am using the node sdk, but i think this is all one in the same and is a service side issue. Delaying or retrying seems to have no effect they always fail with 404

**Comment by arsnyder16** [comment]
Just an update to my comments, so we did some more digging. We are seeing inconsistent behavior between some files from the same users onedrive, even within the same folder. Some files the uploads work fine every time and other file the upload session is created but when you try to use the upload url you get a 404 "ItemNotFound" response.  

here are some details from the failed uploads

Upload Session creation:
`client-request-id: c8ed8467-8893-7f07-3489-d04f76e113be`
`request-id: e3200d76-a160-4630-ac63-05179924df92`

Failed PUT on the uploadUrl
`request-id: f671369f-1021-a000-7eb6-32b87c2884b7`
`sprequestguid: f671369f-1021-a000-7eb6-32b87c2884b7`
`guid in upload url query params: '9091e061-fda2-4d02-adb1-958a1e47cc60'`
Response:
`{"error":{"code":"itemNotFound","message":"The resource could not be found."}}`

Are there any more details that i can supply?

**Comment by arsnyder16** [comment]
We did some more tracking, looks like this issue is as simply as, sharing a file from a folder that is not also shared. 

@darrelmiller Any insight you can supply on the state of the service bugs that are tags this repo? Has anyone tried or successfully reproduced these scenarios on the service side? Has there been acknowledgment from the team?  

Is there a commitment from Microsoft to maintain or add new features to the Graph API? From what i can tell across different related repos and documentation the service side seems stale. Some endpoint have been in beta for years and there are still features lacking from the graphs api that existed in share point previously. Discard checkout is one example.

**Comment by ankku** [comment]
I am trying to upload a file in a folder in SharePoint Library and I am creating the folder before calling the api to upload. I am doing all this by calling graph api in powershell. And i get error as JSON payload is empty. Can someone tell me if thats the bug we are talking here , or is it something with my code??
